### ISSUE TITLE
Add 'unblocked_at' field for Task Serialzer

### ISSUE THREAD

**Issue by bmbouter** [pulp/pulpcore#5057]
**Is your feature request related to a problem? Please describe.**
When a task is in the waiting state, I can't tell if it's waiting due to being blocked by tasks in front of it, e.g. two tasks operating on one repository, or if it's blocked due to a shortage of workers.

**Describe the solution you'd like**
Add a `unblocked_at` field to [the Task serializer](https://docs.pulpproject.org/pulpcore/restapi.html#tag/Tasks/operation/tasks_read).

**Describe alternatives you've considered**
We considered replacing the `waiting` state with a `blocked` and `unblocked` state, but this would be a backwards incompatible change so we can't do it. To get around that backwards incompatible change, we considered adding a new state field altogether, e.g. `new_states` which would use the new states and leaving the old ones alone.

Another alternative was to not put this into Pulp itself, but instead analyze the task data outside of Pulp using Prometheus for example. This argument for doing it in Pulp is that it creates more value for more folks without needing to use external tools.

**Additional context**
This additional data would then be included in the OTEL metrics emitted with task https://github.com/pulp/pulpcore/issues/3821

**Comment by bmbouter** [comment]
@dkliban @decko @mdellweg @lubosmj @dralley FYI ^

**Comment by lubosmj** [comment]
To add more context, this field will contain a timestamp, the same way we have a timestamp for the `finished_at` field, right?

**Comment by decko** [comment]
> To add more context, this field will contain a timestamp, the same way we have a timestamp for the `finished_at` field, right?

Like `blocked_at` / `unblocked_at` @lubosmj?

**Comment by mdellweg** [comment]
> We considered replacing the `waiting` state with a `blocked` and `unblocked` state

Actually, we would split "waiting" into "blocked" and "waiting". I'm still unsure how much we would kick existing client implementations out of orbit, because we would not touch the final states.

When introducing a new timestamp field, I highly recommend moving the whole task state machine into timestamps.
(This is doable, because each state can only ever be entered once. A task cannot be put back into the queue.)

**Comment by ggainey** [comment]
> > We considered replacing the `waiting` state with a `blocked` and `unblocked` state
> 
> Actually, we would split "waiting" into "blocked" and "waiting". I'm still unsure how much we would kick existing client implementations out of orbit, because we would not touch the final states.

Clients that look for "not running|completed|failed" would not notice, anything that waits "until not waiting" would break if we removed the state. Also, "waiting" is what the task is doing, blocked/unblocked tells the user **why** it's waiting. I'd leave "waiting", and set unblocked_at as a condition/addition that is only set on waiting-tasks.

Let me think out loud, just to make sure I'm grokking the change: a task is dispatched, whose required resources are locked. At this moment, it's waiting/blocked. Its required-resource becomes available, but all the workers are in-use, it becomes waiting/unblocked. I am...not sure it's *possible* for a task to move from waiting/unblocked, to waiting/blocked again, given the task-queue? 

So if there is just `unblocked_at`, the waiting/blocked state is "waiting, unblocked_at==None". So you can count "wait for resource-contention" by counting "waiting and not unblocked_at", and count "wait for worker-contention" with "waiting  and unblocked_at".

**Comment by bmbouter** [comment]
> So if there is just `unblocked_at`, the waiting/blocked state is "waiting, unblocked_at==None". So you can count "wait for resource-contention" by counting "waiting and not unblocked_at", and count "wait for worker-contention" with "waiting and unblocked_at".

+1 to :arrow_up: because it is the simplest thing we can do to accomplish the needs we have.

In terms of adding additional timestamp fields for other state transition times, that's fine with me, but can we separate the delivery of that into a followup PR? My goal is to break down the work into the smallest chunk possible and really I only need that 1 field, although I'm supportive of adding more separately.

**Comment by mdellweg** [comment]
> I am...not sure it's _possible_ for a task to move from waiting/unblocked, to waiting/blocked again, given the task-queue?

That's the beauty of it, it impossible. That is the axiom no 1 why the algorithm of the task queue works in the first place.

Me thinking out loud:
The more challenging part of this change is, to find the proper implementation to mark tasks unblocked. A task _always_ needs to enter the task-Q blocked (unless absolutely no resources are required), because anyone can sneak in a task right before that in a racy way. Now there are three situations, a task can become unstuck.
1. Right after it was added. At this time, all workers may be busy.
2. When a worker finished a task. At this time, at least one worker is looking at the task queue.
3. When an immediate task was performed in the api worker. At this time, all workers may be busy.
This brings me right to the conclusion, that we should implement the scan for unblocked tasks in the worker and connect it with pulp worker wakeup both in the sleeping and in the supervise task loop.

All of this should be TDU compatible.

BTW:
Are we sufficiently sure that newly created tasks always appear properly ordered by their creation time? Are these times synced with the end of the transaction? Do we need to add a column lock to task creation? Is there a notion of db wall clock for this? What about clock screw?

**Comment by ggainey** [comment]
> Are we sufficiently sure that newly created tasks always appear properly ordered by their creation time?

[`auto_now_add`](https://docs.djangoproject.com/en/5.0/ref/models/fields/#django.db.models.DateField.auto_now_add) is the object's creation-time, not the transaction-completed-time. Is it possible for a task to have a created-time "earlier" than a task that gets inserted into the queue first (transaction completed first)? Almost certainly. This is the/a downside to how django does so much on the python-side of objects, rather than on the db-side.

The question is, do we, **can** we, care? Tasks should only care about getting dispatched in the order they appear in the queue, and should also only care that _created is equal-or-earlier-than any other timestamp in the task. If two threads create tasks "at the same time", there's no way to guarantee a queue-order - there's too much between "POST" to the API, and "transaction committed".  Anything that requires this level of guarantee, needs to happen in one thread, in order.

**Comment by mdellweg** [comment]
> The question is, do we, **can** we, care?

Yes, we do. Because `created_at` __is__ our queue order.

**Comment by ggainey** [comment]
> > The question is, do we, **can** we, care?
> 
> Yes, we do. Because `created_at` **is** our queue order.

Ah right, of course.  

Well, let me think that through. Thread-A and Thread-B run a dispatch() "at the same time", that requires the same resource. Both create task-objects at the same time, and save() at the same time. If the algorithm to decide "can I mark this as unblocked?" runs *inside that transaction*, then it would be possible for A and B to look at the task queue and decide "I am unblocked" and fill in unblocked_at, then commit. Only one of them will run, and the other will wait - but whoever "loses" will (incorrectly) be marked as unblocked.

What this says to me, is "is this task unblocked" can't be decided as part of dispatch() - it has to be part of "who runs next?"
